src/phoenix/server/api/dataloaders/document_retrieval_metrics.py

```diff
@@ -0,0 +1,93 @@
+from collections import defaultdict
+from itertools import groupby
+from typing import (
+    AsyncContextManager,
+    Callable,
+    DefaultDict,
+    Dict,
+    List,
+    Optional,
+    Set,
+    Tuple,
+)
+
+import numpy as np
+from sqlalchemy import select
+from sqlalchemy.ext.asyncio import AsyncSession
+from strawberry.dataloader import DataLoader
+from typing_extensions import TypeAlias
+
+from phoenix.db import models
+from phoenix.metrics.retrieval_metrics import RetrievalMetrics
+from phoenix.server.api.types.DocumentRetrievalMetrics import DocumentRetrievalMetrics
+
+RowId: TypeAlias = int
+NumDocs: TypeAlias = int
+EvalName: TypeAlias = Optional[str]
+Key: TypeAlias = Tuple[RowId, NumDocs, EvalName]
+
+
+class DocumentRetrievalMetricsDataLoader(DataLoader[Key, List[DocumentRetrievalMetrics]]):
+    def __init__(self, db: Callable[[], AsyncContextManager[AsyncSession]]) -> None:
+        super().__init__(load_fn=self._load_fn)
+        self._db = db
+
+    async def _load_fn(self, keys: List[Key]) -> List[List[DocumentRetrievalMetrics]]:
+        results: Dict[Key, List[DocumentRetrievalMetrics]] = {key: [] for key in keys}
+        requests: DefaultDict[RowId, DefaultDict[NumDocs, Set[EvalName]]] = defaultdict(
```

Comment:
I think I would have an easier time understanding the loop at the end if we made this into a dictionary of type `Dict[Tuple[RowId, EvalName], NumDocs]`. This would avoid the extra level of nesting in the loop at the end.